[SYCL][DOC] Add extension for FPGA kernel interface properties

tiwaria1 opened the pull request:
## SYCL extension contains the following new kernel properties
- `streaming_interface<...>`
- `register_map_interface<...>`
The first two properties take enum arguments that provide the compiler information about whether the logic downstream to the kernel will back-pressure the kernel or not.
- `pipelined<N>`
Takes an integer, non-zero values specify minimum cycles between kernel invocations, and 0 specifies that pipelining should be disabled.

tiwaria1 commented on sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:
```diff
@@ -0,0 +1,198 @@
+This extension provides a feature-test macro as described in the core SYCL
+specification section 6.3.3 "Feature test macros".  Therefore, an
+implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_KERNEL_PROPERTIES` to one of the values defined in the table
```
```suggestion
`SYCL_EXT_INTEL_FPGA_KERNEL_INTERFACE_PROPERTIES` to one of the values defined in the table
```

GarveyJoe replied:
```suggestion
`SYCL_EXT_INTEL_FPGA_KERNEL_INTERFACE_PROPERTIES` to one of the values defined in the table
```

GarveyJoe commented on sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:
```diff
@@ -0,0 +1,198 @@
+
+== Overview
+
+This extension specifies the kernel properties designed for kernels that target
```
This sentence is simultaneously too specific (as the "the" implies that no other extensions might add FPGA kernel properties) and too vague (there's no indication of what these properties are about).  I suggest something like this instead:

This extension introduces kernel properties to specify how or when control and data signals can be passed in to or out of an FPGA kernel.  These properties are meaningless on non-FPGA devices and can be ignored on such devices.

Given that there are only two properties, you could even go further and summarize them both like so:

The ip_interface_streaming property requires the kernel to have a dedicated output port for each control or data signal, as opposed to using a shared CSR interface.  The ip_pipelined property directs the compiler to pipeline calls to the kernel so that multiple invocations of the kernel can be in flight simultaneously.

tiwaria1 replied:
Updated. Since it's 4 properties now, I have placed the additional description in the properties table.

GarveyJoe commented on sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:
```diff
@@ -0,0 +1,198 @@
+ For convenience, a constant bool named `stall_free_return` with the value
+ `true` is provided.
+
+|`ip_pipelined`
```
I'm not a big fan of the "ip" at the beginning of each of these names.  What concept are you trying to get across with this prefix?

tiwaria1 replied:
Changed the names to:
```
streaming_kernel
csr_kernel
stall_free_return_kernel
pipelined_kernel
```
1. Perhaps `stall_free_return_kernel` should be renamed to `no_backpressure_kernel`
2. Other kernel properties do not have a '_kernel' post-fix. I think we should instead use the names:
```
streaming_interface
csr_interface
no_backpressure_interface
pipelined
```
Thoughts?

GarveyJoe replied:
#2: agreed.  Let's be consistent with other kernel properties.
#1: let's run it by internal users to see whether they prefer "stall_free" or "no_backpressure"

tiwaria1 replied:
Updated based on the discussion with internal users.

GarveyJoe commented on sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:
```diff
@@ -0,0 +1,198 @@
+== Overview
+
+This extension specifies the kernel properties designed for kernels that target
+the FPGA device. These properties are supported by the Intel FPGA backend
```
Which compilers support this extension is irrelevant for the sake of the definition of the extension itself.

tiwaria1 replied:
Updated.

GarveyJoe commented on sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:
```diff
@@ -0,0 +1,198 @@
+|`ip_interface_streaming`
+|The `ip_interface_streaming` property adds the requirement that the kernel must
+ have dedicated ports for input / output signals in the generated
+ Register Transfer Level (RTL) module.
```
I would drop the mention of RTL.  It confuses things to refer to the description of the hardware as being separate from the hardware itself.  
Also, I think your intent is for this to apply to both control and data signals.  Can you state the explicitly?  And assuming that this should apply to data signals as well, should it apply to all inputs/outputs from the kernel, i.e. pipes as well, or only kernel arguments?

tiwaria1 replied:
Added explicit statement. Specified that it should apply to kernel arguments and control flow signals only. Including pipes seems to make sense. I need to think a bit more about including them.

GarveyJoe commented on sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:
```diff
@@ -0,0 +1,198 @@
+ 
+ The property takes a constant boolean template argument. When the value of the
+ argument is `true` it adds the requirement that the kernel module should not
+ take a 'stall' control flow signal from downstream logic as input. The kernel
```
Why is the stall-freedom a template parameter of ip_interface_streaming instead of its own property?  How can a user guarantee that done signals are not lost in a pipelined kernel with a CSR interface if there is no way to backpressure the done signal?

tiwaria1 replied:
It should be its own property. Updated.

tiwaria1 commented on sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:
```diff
@@ -0,0 +1,198 @@
+used by permission by Khronos.
+
+This extension introduces a replacement for the kernel attributes defined in
+Section 5.8.1 of the SYCL 2020 specification, in the form of a property list
```
Updated.

tiwaria1 commented on sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:
```diff
@@ -0,0 +1,198 @@
+This extension is written against the SYCL 2020 specification, Revision 4 and
+the following extensions:
+
+- link:sycl_ext_oneapi_properties.asciidoc[sycl_ext_oneapi_properties]
```
I removed the hyperlink for now. Need to figure out how to link to a file in a different directory.

tiwaria1 replied:
Corrected the link.

keryell commented on sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:
```diff
@@ -0,0 +1,213 @@
+    *a = *b + *c;
+  }
+
+  auto get(sycl::ext::oneapi::experimental::properties_tag) {
```
Is this a magic function that encode the properties? How it is connected to the `single_task` below?

keryell commented on sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:
```diff
@@ -0,0 +1,213 @@
+|`pipelined_kernel`
+| The `pipelined_kernel` property directs the compiler to pipeline calls to the
+ kernel so that multiple invocations of the kernel can be in flight
+ simultaneously.
```
Do you mean several instances of a `single_task` or this is pipelining at the work-item level?

GarveyJoe replied:
This refers to several instances of a kernel, most commonly a single_task kernel.  It does not refer to pipelining of work-items.  However, there's no reason this couldn't be applied to a parallel_for kernel to request that successive launches of that kernel be pipelined _in addition to_ the usual work item pipelining.  In such a kernel, the work items could be injected into the datapath one after another and, after the last work item from a given kernel invocation has been injected, the first work-item of the next invocation could be injected without waiting for the pipeline to drain.  

I don't see much value in a control to enable or disable pipelining at the work-item level because that's nearly always beneficial.  This invocation pipelining is more situational though as it only provides value if the user intends to launch the same kernel multiple times.

keryell commented on sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:
```diff
@@ -0,0 +1,213 @@
+  sycl::ext::oneapi::experimental::streaming_kernel,
+  sycl::ext::oneapi::experimental::stall_free_return_kernel};
+
+q.single_task(properties, [=]() {
```
```suggestion
q.single_task(properties, [=] {
```

GarveyJoe commented on sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:
```diff
@@ -0,0 +1,296 @@
+ its input / output control and kernel argument data signals placed in a shared
+ Control and Status Register (CSR) map. The following values are supported:
+
+ * `wait_for_done_read`: Provides the compiler the guarantee that downstream
```
I thought we didn't want to use a read-on-clear approach and had instead discussed requiring the user to overwrite the done bit with a 0 if they want backpressure.  If we go that way, we should probably call this "wait_for_done_write" or "wait_for_done_clear".

GarveyJoe replied:
Also, this definition is vague and misleading.  The user is guaranteeing anything to the compiler, they are asking us to build logic that backpressures the kernel until they've told us they've seen the done.

GarveyJoe commented on sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:
```diff
@@ -0,0 +1,296 @@
+ * `wait_for_done_read`: Provides the compiler the guarantee that downstream
+    logic will back-pressure the kernel.
+
+ * `do_not_wait_for_done_read`: Provides the compiler the guarantee that
```
This one is more of a guarantee but it still seems like the wrong choice of words.  The user is telling the compiler that it can build hardware that doesn't backpressure the kernel.  I.e. that the user's system is either tolerant to dropped dones or that they are going to throttle the kernel themselves sufficiently to ensure that no done are dropped.

GarveyJoe commented on sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:
```diff
@@ -0,0 +1,296 @@
+ have dedicated ports for input / output signals. This applies for both
+ control, and kernel argument data signals. The following values are supported:
+
+ * `accept_downstream_stall`: Provides the compiler the guarantee that
```
A "guarantee that downstream logic will back-pressure the kernel" sounds like a guarantee that the ready signal is never high and thus that the kernel never needs to do anything.  That's definitely not what you intended.  You could reword this as: "Instructs the compiler to build a backpressure input port to the kernel (e.g. a ready signal for the avalon-ST protocol)."

GarveyJoe commented on sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:
```diff
@@ -0,0 +1,296 @@
+ * `accept_downstream_stall`: Provides the compiler the guarantee that
+    downstream logic will back-pressure the kernel.
+
+ * `remove_downstream_stall`: Provides the compiler the guarantee that
```
It goes beyond that, right?  This directs the compiler to omit the port entirely.

GarveyJoe commented on sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:
```diff
@@ -0,0 +1,296 @@
+
+ When the boolean parameter is set to `true`, the property directs the compiler
+ to pipeline calls to the kernel such that multiple invocations of the kernel
+ can be in flight simultaneously. When the boolean paramter is set to `false`,
```
```suggestion
 can be in flight simultaneously. When the boolean parameter is set to `false`,
```

GarveyJoe commented on sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:
```diff
@@ -0,0 +1,296 @@
+ can be in flight simultaneously. When the boolean paramter is set to `false`,
+ the compiler will not pipeline kernel invocations.
+
+ The second property argument should be used to specify how many cycles can
```
```suggestion
 The second property argument specifies how many cycles can
```

GarveyJoe commented on sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:
```diff
@@ -0,0 +1,296 @@
+ the compiler will not pipeline kernel invocations.
+
+ The second property argument should be used to specify how many cycles can
+ exist between multiple invocations of the kernel. The `initiation_interval`
```
```suggestion
 exist between successive invocations of the kernel. The `initiation_interval`
```

GarveyJoe commented on sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:
```diff
@@ -0,0 +1,296 @@
+ interface is shown below.
+
+The example assumes that the kernel will not be back-pressured by the downstream
+logic and hence uses the property `stall_free_return_kernel`:
```
This is the old name and also the example below does accept a stall so these need to be made consistent.

GarveyJoe commented on sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:
````diff
@@ -0,0 +1,296 @@
+```
+
+The example below shows how to launch a pipelined kernel with a streaming
+interface, and with every kernel invocation being launched every 2 cycles.
````
```suggestion
interface, and with a new kernel invocation being launched every 2 cycles.
```

keryell commented on sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:
````diff
@@ -0,0 +1,296 @@
+== Kernel Properties
+
+```c++
+namespace sycl {
+namespace ext {
+namespace oneapi {
+namespace experimental {
````
C++17 is good for FPGA :-)
```suggestion
namespace sycl::ext::oneapi::experimental {
```
Along this, it should be good to have at least one example showing how a namespace alias like
```c++
namespace oe = sycl::ext::oneapi::experimental;
```
can be used to write shorter code, just in case some HLS people without any SYCL or C++ knowledge can see how to write compact code and not think that SYCL is overly verbose.

tiwaria1 commented:
ping @intel/dpcpp-specification-reviewers

steffenlarsen commented on sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:
```diff
@@ -0,0 +1,302 @@
+
+|`pipelined`
+|Two sets of values are accepted as property parameters - a boolean, and another
+ property `initiation_interval`.
```
Is there a benefit to defining `initiation_interval` as a property? I worry that having it use the property interface may wrongfully give the user the idea that it is to be used like a property, while it is simply intended for use inside the `pipelined` property. It also adds more to the implementation, as properties need to define a number of things under the hood, whereas if we define it as some simple wrapper class we can avoid that.

tiwaria1 replied:
Hi Steffen, I have updated the spec, please review again. Thank you!

steffenlarsen commented on sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:
```diff
@@ -96,10 +96,9 @@ enum class register_map_interface_options_enum {
-    std::integral_constant<int, initiation_interval_value>>;
+template <int N>
+struct initiation_interval {
+  int initiation_interval_value = N;
```
Is this expected to mutate? Otherwise it would be better to have it be `static constexpr`. Also, the naming of the value member seems a little verbose, given it simply repeats the type name. Would `value` be too simple a name?

tiwaria1 replied:
Value is not expected to mutate so I will make it a `static constexpr` and yes I can rename the field to `value` as I don't think that the more verbose name is needed.

steffenlarsen reviewed (approved):
LGTM!

GarveyJoe commented on sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:
```diff
@@ -0,0 +1,296 @@
+== Overview
+
+This extension introduces kernel properties to specify how or when control and
+data signals can be passed in to or out of an FPGA kernel. These properties are
```
```suggestion
data signals can be passed into or out of an FPGA kernel. These properties are
```

GarveyJoe commented on sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:
```diff
@@ -0,0 +1,296 @@
+ * `wait_for_done_write`: Directs the compiler to generate logic that
+    back-pressures the kernel until the kernel is notified that its completion
+    has been detected. The kernel will be notified when the register it writes
+    its completion signal to, is set to 1.
```
```suggestion
    its completion signal to is set to 1.
```

GarveyJoe commented on sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:
```diff
@@ -0,0 +1,296 @@
+   invocations such that an invocation is launched every N cycles.
+ 
+ If the `initiation_interval` argument is not specified, the compiler will
+ attempt to launch the kernel invocations as fast as it can.
```
I propose we leave this as an implementation detail.  As written it implies that an implementation must aggressively pursue II=1 even if that doesn't make sense (e.g. because the kernel contains a loop with II=1 with high trip count which will be the bottleneck anyways).

tiwaria1 replied:
Updated to mark this as implementation defined.

keryell reviewed (commented):
`pipelined` could be simplified to take a single parameter with value `false`, `true` or some II value.
The current cartesian product with 2 parameters is useless since `false` and an II is meaningless.

keryell commented:
Also having a default `true` value for the `pipelined` when used seems an interesting shortcut.

GarveyJoe commented on sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:
```diff
@@ -0,0 +1,296 @@
+ * `wait_for_done_write`: Directs the compiler to generate logic that
+    back-pressures the kernel until the kernel is notified that its completion
+    has been detected. The kernel will be notified when the register it writes
+    its completion signal to is set to 1.
```
```suggestion
    its completion signal to is set to 0.
```

If it's a "done" register then the kernel should write 1 to it when it finishes and the user should write 0 to it to say "I've seen your done".

GarveyJoe commented on sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:
```diff
@@ -0,0 +1,296 @@
+ 
+ If the `initiation_interval` argument is not specified, the behavior is
+ implementation defined.
+ 
```
You might want to mention here that if the bool is false, the initiation_interval parameter is meaningless.

GarveyJoe commented on sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:
```diff
@@ -0,0 +1,296 @@
+the following extensions:
+
+- link:../experimental/sycl_ext_oneapi_properties.asciidoc[sycl_ext_oneapi_properties]
+- link:sycl_ext_oneapi_kernel_properties.asciidoc[sycl_ext_oneapi_kernel_properties]
```
Links are broken

tiwaria1 replied:
They seem to be working now.
oneapi properties is in experimental folder: https://github.com/intel/llvm/tree/sycl/sycl/doc/extensions/experimental
and oneapi kernel properties will be in the proposed folder after 5755 PR goes in : https://github.com/intel/llvm/pull/5755/files

tiwaria1 commented:
@steffenlarsen I addressed the remaining comments. Please review and merge this PR. Based on your namespace comment on a different PR I have updated the namespace here to be `sycl::ext::intel::experimental` since this is a FPGA specific extension.

tiwaria1 commented on sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:
```diff
@@ -0,0 +1,287 @@
+
+struct register_map_interface_key {
+  template <register_map_interface_options_enum option>
+  using value_t = property_value<register_map_interface_key,
```
Hi @steffenlarsen , Instead of `property_value`, should this be rewritten as `sycl::ext::oneapi::properties::property_value` ? Or perhaps I should add a `using sycl::ext::oneapi::properties` at the top?

tiwaria1 replied:
Updated!

steffenlarsen commented on sycl/doc/extensions/proposed/sycl_ext_intel_fpga_kernel_interface_properties.asciidoc:
```diff
@@ -0,0 +1,287 @@
+using sycl::ext::intel::experimental;
+{
+  ...
+  properties kernel_properties{streaming_interface_remove_downstream_stall};
```
```suggestion
  auto kernel_properties = properties{streaming_interface_remove_downstream_stall};
```
And similar in other places.

keryell replied:
This is bike-shedding but I usually push into the other direction, when `auto` is useless and longer. :-)

steffenlarsen replied:
Honestly, I wasn't sure it would accept the type without the template arguments explicitly there, but I guess I've learned something new. In that case I definitely like the old version better. Thanks @keryell! 😄

steffenlarsen reviewed (commented):
Just a few more comments, then I think I'm happy!

tiwaria1 commented:
> Just a few more comments, then I think I'm happy!

Hey Steffen, I addressed the other comments. Please take a look.

steffenlarsen reviewed (approved):
LGTM!
